pennylane/interfaces/jacobian_products.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Defines classes that take the vjps, jvps, and jacobians of circuits.
+"""
+import abc
+from functools import partial
+from typing import Tuple, Callable, Optional
+
+import pennylane as qml
+from pennylane.tape import QuantumScript
+from pennylane.typing import ResultBatch, TensorLike
+
+Batch = Tuple[QuantumScript]
+
+
+class JacobianProductCalculator(abc.ABC):
+    """Provides methods for calculating the JVP/VJP between the Jacobians of tapes and tangents/cotangents."""
+
+    @abc.abstractmethod
+    def execute_and_compute_jvp(
+        self, tapes: Batch, tangents: Tuple[Tuple[TensorLike]]
+    ) -> Tuple[ResultBatch, Tuple]:
+        """Calculate both the results for a batch of tapes and the jvp.
+
+        This method is required to compute JVPs in the JAX interface.
+
+        Args:
+            tapes (tuple[`~.QuantumScript`]): The batch of tapes to take the derivatives of
+            tangents (Sequence[Sequence[TensorLike]]): the tangents for the parameters of the tape.
+                The ``i``th tangent corresponds to the ``i``th tape, and the ``j``th entry into a
+                tangent entry corresponds to the ``j``th trainable parameter of the tape.
+
+        Returns:
+            ResultBatch, TensorLike: the results of the execution and the jacobian vector product
+
+        **Examples:**
+
+        >>> tape0 = qml.tape.QuantumScript([qml.RX(0.1, wires=0)], [qml.expval(qml.PauliZ(0))])
+        >>> tape1 = qml.tape.QuantumScript([qml.RY(0.2, wires=0)], [qml.expval(qml.PauliZ(0))])
+        >>> batch = (tape0, tape1)
+        >>> tangents0 = (1.5, )
+        >>> tangents1 = (2.0, )
+        >>> tangents = (tangents0, tangents1)
+        >>> results, jvps = jp_method.execute_and_compute_jvp(batch, tangents)
+        >>> expected_results = (np.cos(0.1), np.cos(0.2))
+        >>> qml.math.allclose(results, expected_results)
+        True
+        >>> jvps
+        (array(-0.14975012), array(-0.39733866))
+        >>> expected_jvps = 1.5 * -np.sin(0.1), 2.0 * -np.sin(0.2)
+        >>> qml.math.allclose(jvps, expected_jvps)
+        True
+
+        """
```

Comment:
Just because this has been causing problems in the old interface functions: Are tape parameters assumed to be scalars? If so, could we state this explicitly here? :)

Reply:
If neither `batch_jvp` or adjoint differentation support it, I'm not sure its in the scope of this PR to add.  I would be willing to break off another task for later investigation.  These classes are basically to serve as replacement for logic like:

https://github.com/PennyLaneAI/pennylane/blob/67b92d276727eab070d85744eae90e88386eeaae/pennylane/interfaces/autograd.py#L492

It's moving code from that nested if-elif-elif block to simply `compute_vjp`.

Reply:
Alright, I see :) 
Fine by me to separate out the addition of support comments to another PR.